staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go

```diff
@@ -0,0 +1,1459 @@
+*/
+
+package filters
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"net/http/httptrace"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+func TestPerRequestWriteDeadlineWithWriteShouldReturnErrorAfterTimeout(t *testing.T) {
+	// This test documents the behavior of the per request write deadline
+	// using a standard net http server.
+	//
+	// scenario: we want to observe the circumstances under which the Write
+	// method of the ResponseWriter object returns an "i/o timeout" error:
+	//  - a) write deadline is set to 100ms
+	//  - b) the handler waits until the write deadline exceeds
+	//  - c) after timeout occurs, the handler keeps writing 1kB of data at a time
+	// we expect the following:
+	//  - d) client: expected to receive an error from the server
+	//  - e) server: the Write method of the ResponseWriter object should
+	//       return an "i/o timeout" error once its internal buffer is full
+	t.Parallel()
+
+	const deadline = 100 * time.Millisecond
+	tests := []struct {
+		protoMajor         int
+		waiter             waiter
+		clientErrVerifier  verifier
+		handlerErrVerifier verifier
+	}{
+		{
+			protoMajor: 1, // http/1x
+			// write timeout is set to 100ms, a wait of 5s should be
+			// enough to withstand flakes in CI.
+			waiter:             &waitWithDuration{after: 5 * time.Second},
+			clientErrVerifier:  wantContains{contains: "local error: tls: bad record MAC"},
+			handlerErrVerifier: wantError{err: os.ErrDeadlineExceeded},
+		},
+		{
+			protoMajor:         2, // http/2.0
+			waiter:             &waitWithChannelClose{after: make(chan time.Time)},
+			clientErrVerifier:  wantContains{"stream error: stream ID 1; INTERNAL_ERROR; received from peer"},
+			handlerErrVerifier: wantError{err: os.ErrDeadlineExceeded},
+		},
+	}
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("HTTP/%d", test.protoMajor), func(t *testing.T) {
+			handlerErrCh := make(chanErr, 1)
+			server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+				defer close(handlerErrCh)
+				if req.ProtoMajor != test.protoMajor {
+					t.Errorf("server: expected ProtoMajor: %d, but got: %s", test.protoMajor, req.Proto)
+					return
+				}
+
+				// a) set write deadline for this request
+				ctrl := http.NewResponseController(w)
+				if err := ctrl.SetWriteDeadline(time.Now().Add(deadline)); err != nil {
+					t.Errorf("server: unexpected error from SetWriteDeadline: %v", err)
+					return
+				}
+
+				// b) wait until the write deadline exceeds
+				<-test.waiter.wait()
+
+				// c) keep writing 1kB of data at a time until
+				// Write returns an error
+				func() {
+					now := time.Now()
+					written := 0
+					defer func() {
+						t.Logf("server: after timeout, Write wrote %dkB, duration: %s", written/1000, time.Since(now))
+					}()
+					for {
+						n, err := w.Write(bytes.Repeat([]byte("a"), 1000))
```

Comment:
shouldn't we check that none of these makes it to the client? You claim something below before flushing.